### PR TITLE
Distributed singleton should be more careful with handover

### PR DESCRIPTION
**Pull request by ktoso**
Don't review yet;

Resolves https://github.com/apple/swift-distributed-actors/issues/1123 by hardedning how the redirecring is done. This is still not optimal but hit IRGen issue

### CODE REVIEW

File: Sources/DistributedCluster/Plugins/ClusterSingleton/ClusterSingletonBoss.swift [hunk 1]
```diff
@@ -155,50 +161,87 @@ internal distributed actor ClusterSingletonBoss<Act: ClusterSingleton>: ClusterS
...
+                maxAttempts: settings.locateActiveSingletonBackoff.maxAttempts
+            ).attempt {
+                // confirm tha the boss is hosting the singleton, if not we may have to wait and try again
+                do {
+                    guard ((try? await targetSingletonBoss.hasActiveSingleton()) ?? false) else {
+                        throw SingletonNotFoundNoExpectedNode(id: self.settings.name, node)
+                    }
+                } catch {
+                    throw error
+                }
```
Comment (ktoso) [inline Sources/DistributedCluster/Plugins/ClusterSingleton/ClusterSingletonBoss.swift:216]: This do/catch is just here so we can add debug logging, will do shortly

File: Sources/DistributedCluster/Plugins/ClusterSingleton/ClusterSingletonBoss.swift [hunk 2]
```diff
@@ -155,50 +161,87 @@ internal distributed actor ClusterSingletonBoss<Act: ClusterSingleton>: ClusterS
...
+            targetSingletonBossID.path = self.id.path
+            let targetSingletonBoss = try Self.resolve(id: targetSingletonBossID, using: self.actorSystem)
+
+            let targetSingleton: Act = try await Backoff.exponential(
+                initialInterval: settings.locateActiveSingletonBackoff.initialInterval,
+                multiplier: settings.locateActiveSingletonBackoff.multiplier,
+                capInterval: settings.locateActiveSingletonBackoff.capInterval,
+                randomFactor: settings.locateActiveSingletonBackoff.randomFactor,
+                maxAttempts: settings.locateActiveSingletonBackoff.maxAttempts
+            ).attempt {
```
Comment (ktoso) [inline Sources/DistributedCluster/Plugins/ClusterSingleton/ClusterSingletonBoss.swift:208]: `Backoff.exponential().attemtpt {}` is very nice, we're finally beginning to reap benefits of async/await inside the cluster :)

**Comment by ktoso** [comment]
Once again a different flaky one: https://github.com/apple/swift-distributed-actors/issues/1124
And https://github.com/apple/swift-distributed-actors/issues/1128 -- this seems new but also only in one of the runs, so adding to the flaky list to investigate

This PR is to resolve a different flaky issue so I'll merge

**Comment by ktoso** [comment]
Minor follow up ticket https://github.com/apple/swift-distributed-actors/issues/1129

**Comment by ktoso** [comment]
I'll merge so we have less flaky tests but please do review when you're back from break @yim-lee !

File: Sources/DistributedCluster/Backoff.swift [hunk 3]
```diff
@@ -208,6 +206,38 @@ public struct ExponentialBackoffStrategy: BackoffStrategy {
...
+    /// Attempt to execute the passed `operation` at most `maxAttempts` times while applying the expected backoff strategy.
+    ///
+    /// - Parameters:
+    ///   - operation: The operation to run, potentially multiple times until successful or maxAttempts were made
+    /// - Throws: the last error thrown by `operation`
+    public func attempt<Value>(_ operation: () async throws -> Value) async throws -> Value {
+        var backoff = self
+        var lastError: Error?
+        defer {
+            pprint("RETURNING NOW singleton")
```
Comment (yim-lee) [inline Sources/DistributedCluster/Backoff.swift:218]: Not intended to keep here I think?

File: Sources/DistributedCluster/Plugins/ClusterSingleton/ClusterSingletonBoss.swift [hunk 4]
```diff
@@ -155,50 +161,87 @@ internal distributed actor ClusterSingletonBoss<Act: ClusterSingleton>: ClusterS
...
+            self.updateSingleton(targetSingleton)
 
-            let singleton = try Act.resolve(id: targetSingletonID, using: self.actorSystem)
-            self.updateSingleton(singleton)
         case .none:
             self.updateSingleton(nil)
         }
     }
 
+    // FIXME: would like to return `Act?` but can't via the generic combining with Codable: rdar://111664985 & https://github.com/apple/swift/issues/67090
```
Comment (yim-lee) [inline Sources/DistributedCluster/Plugins/ClusterSingleton/ClusterSingletonBoss.swift:231]: 👍

File: Sources/DistributedCluster/Plugins/ClusterSingleton/ClusterSingletonBoss.swift [hunk 5]
```diff
@@ -267,8 +310,8 @@ internal distributed actor ClusterSingletonBoss<Act: ClusterSingleton>: ClusterS
             try await singletonFactory(self.actorSystem)
         }
 
-        await singleton.whenLocal { __secretlyKnownToBeLocal in // TODO(distributed): this is annoying, we must track "known to be local" in
```
Comment (yim-lee) [inline Sources/DistributedCluster/Plugins/ClusterSingleton/ClusterSingletonBoss.swift:270]: ❤️

File: Sources/DistributedCluster/Backoff.swift (hunk 3 above)
Comment (ktoso) [reply Sources/DistributedCluster/Backoff.swift:218]: Thanks!

File: Sources/DistributedCluster/Plugins/ClusterSingleton/ClusterSingletonBoss.swift (hunk 4 above)
Comment (ktoso) [reply Sources/DistributedCluster/Plugins/ClusterSingleton/ClusterSingletonBoss.swift:231]: Solution incoming here: https://github.com/apple/swift/pull/67117

**Pull request merged**